feat(api): add `minio-client` provider

HarshPatel5940 opened the pull request:
### **User description**
## Description

This pull request adds the `minio-client` provider to the project. The `minio-client` allows for interaction with a MinIO server, including uploading, downloading, and deleting files. This feature will enhance the functionality of the project by providing support for MinIO storage.

Fixes #120

## Mentions

_Mention and tag the people_


## Developer's checklist
- [x] My PR follows the style guidelines of this project
- [x] I have performed a self-check on my work

**If changes are made in the code:**

- [x] I have followed the [coding guidelines](https://google.github.io/styleguide/jsguide.html)
- [x] My changes in code generate no new warnings
- [ ] My changes are breaking another fix/feature of the project
- [ ] I have added test cases to show that my feature works
- [ ] I have added relevant screenshots in my PR
- [x] There are no UI/UX issues


### Documentation Update

- [x] This PR requires an update to the documentation at [docs.keyshade.xyz](https://docs.keyshade.xyz)
- [x] I have made the necessary updates to the documentation, or no documentation changes are required.


___

### **PR Type**
Enhancement, Documentation, Configuration changes


___

### **Description**
- Added `MinioProvider` to interact with MinIO server, including functions for uploading, downloading, and deleting files.
- Added MinIO related environment variables to the schema and documented them.
- Registered `MinioProvider` in the provider module.
- Added `minio` package to dependencies.
- Added MinIO service configuration to Docker Compose and test Docker Compose files.
- Added example MinIO environment variables to `.env.example`.


___



### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement
</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>minio.provider.ts</strong><dd><code>Add `MinioProvider` for MinIO server interactions.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

apps/api/src/provider/minio.provider.ts
<li>Added <code>MinioProvider</code> to interact with MinIO server.<br> <li> Implemented functions for uploading, downloading, and deleting files.<br> <li> Included environment variable checks and bucket creation logic.<br>


</details>
    

  </td>
  <td><a href="https://github.com/keyshade-xyz/keyshade/pull/237/files#diff-2767b86ae65d0c0ca424ed8bc2f02e45b452f50277e4c7d04cca7149868b8540">+74/-0</a>&nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>provider.module.ts</strong><dd><code>Register `MinioProvider` in provider module.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

apps/api/src/provider/provider.module.ts
- Registered `MinioProvider` in the provider module.



</details>
    

  </td>
  <td><a href="https://github.com/keyshade-xyz/keyshade/pull/237/files#diff-13afe5e8d9a5fdbcf834ba56d994b254f353fdae3c3a2ee376018a0cbf10b51a">+6/-1</a>&nbsp; &nbsp; &nbsp; </td>
</tr>                    
</table></td></tr><tr><td><strong>Configuration changes
</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>env.schema.ts</strong><dd><code>Add MinIO environment variables to schema.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

apps/api/src/common/env/env.schema.ts
- Added MinIO related environment variables to the schema.



</details>
    

  </td>
  <td><a href="https://github.com/keyshade-xyz/keyshade/pull/237/files#diff-3ac6831dfb43e9b21b1351465f5eebdf99c9e929c9abacb059cbf16f3d0b8c1d">+7/-0</a>&nbsp; &nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>docker-compose.yml</strong><dd><code>Add MinIO service to Docker Compose.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

docker-compose.yml
- Added MinIO service configuration.



</details>
    

  </td>
  <td><a href="https://github.com/keyshade-xyz/keyshade/pull/237/files#diff-e45e45baeda1c1e73482975a664062aa56f20c03dd9d64a827aba57775bed0d3">+13/-0</a>&nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>.env.example</strong><dd><code>Add example MinIO environment variables.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

.env.example
- Added example MinIO environment variables.



</details>
    

  </td>
  <td><a href="https://github.com/keyshade-xyz/keyshade/pull/237/files#diff-a3046da0d15a27e89f2afe639b25748a7ad4d9290af3e7b1b6c1a5533c8f0a8c">+8/-2</a>&nbsp; &nbsp; &nbsp; </td>
</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>docker-compose-test.yml</strong><dd><code>Add MinIO service to test Docker Compose.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

docker-compose-test.yml
- Added MinIO service configuration for testing.



</details>
    

  </td>
  <td><a href="https://github.com/keyshade-xyz/keyshade/pull/237/files#diff-6e72fc85bdf9d7aa207102402a2de7cf7aec4109d2378b66921a5c70d07a4f1f">+13/-0</a>&nbsp; &nbsp; </td>
</tr>                    
</table></td></tr><tr><td><strong>Documentation
</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>environment-variables.md</strong><dd><code>Document MinIO environment variables.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

docs/contributing-to-keyshade/environment-variables.md
- Documented new environment variables for MinIO configuration.



</details>
    

  </td>
  <td><a href="https://github.com/keyshade-xyz/keyshade/pull/237/files#diff-2bf98f3c57ba5baaa3603870cea1ee6984fea12550638d22e65bfe23f3234af1">+9/-1</a>&nbsp; &nbsp; &nbsp; </td>
</tr>                    
</table></td></tr><tr><td><strong>Dependencies
</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>package.json</strong><dd><code>Add `minio` package to dependencies.</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

apps/api/package.json
- Added `minio` package to dependencies.



</details>
    

  </td>
  <td><a href="https://github.com/keyshade-xyz/keyshade/pull/237/files#diff-2c40985d6d91eed8ae85ec1c8e754a85984ee32e156a600d2b7a467423d7e338">+4/-2</a>&nbsp; &nbsp; &nbsp; </td>
</tr>                    
</table></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

rajdip-b commented on apps/api/src/provider/minio.provider.ts:
```diff
@@ -0,0 +1,65 @@
+      isServiceLoaded = true
+    }
+
+    async function uploadFile(file) {
```
Functions that communicate with the client shoud always check `isServiceLoaded` before making some function. Since this is a direct dependency, the best thing would be to throw a `500` error.

HarshPatel5940 replied:
so you would like me to return a 500 error? along with message stating minio has not been loaded?

HarshPatel5940 replied:
```ts 
   if (!isServiceLoaded) {
        return new HttpException('Minio Client has not loaded', 500)
      }
```
in start of all the functions?

---

would it be better if we just exit the program with (1) during initial boot process

rajdip-b commented on docker-compose.yml:
```diff
@@ -19,6 +20,22 @@ services:
   keyshade-dev:
     driver: bridge
+
+volumes:
```
I think you can skip the volume creation

HarshPatel5940 replied:
mhmm, ok.

rajdip-b commented on docker-compose.yml:
```diff
@@ -19,6 +20,22 @@ services:
+      - '9000:9000'
+      - '9001:9001'
+    volumes:
+      - 'minio_data:/data'
```
The data would be stored under `/data/minio`. Refer the other two containers to see how.

HarshPatel5940 replied:
noted

rajdip-b commented on docker-compose.yml:
We would need a similar modification in the test compose file aswell.

HarshPatel5940 replied:
noted.

rajdip-b commented on apps/api/src/provider/minio.provider.ts:
```diff
@@ -0,0 +1,74 @@
+      return
+    }
+
+    const minioClient: Minio.Client = new Minio.Client({
```
Use try-catch in here because it might run into permission errors. If you face any error, make sure the application exits with a non-zero status code. Make sure you emit the error log.

HarshPatel5940 replied:
noted.

rajdip-b commented on apps/api/src/provider/minio.provider.ts:
```diff
@@ -0,0 +1,74 @@
+    })
+
+    const bucketName = process.env.MINIO_BUCKET_NAME
+    createBucketIfNotExists()
```
this should be awaited

HarshPatel5940 replied:
ok

rajdip-b commented on apps/api/src/provider/minio.provider.ts:
```diff
@@ -0,0 +1,74 @@
+        return new InternalServerErrorException('Minio Client has not loaded')
+      }
+      const fileName = `${Date.now()}-${file.originalname}`
+      await minioClient.putObject(bucketName, fileName, file.buffer, file.size)
```
Use try catch

rajdip-b commented on apps/api/src/provider/minio.provider.ts:
```diff
@@ -0,0 +1,74 @@
+      if (!isServiceLoaded) {
+        return new InternalServerErrorException('Minio Client has not loaded')
+      }
+      return await minioClient.presignedUrl('GET', bucketName, fileName)
```
Use try catch

rajdip-b commented on apps/api/src/provider/minio.provider.ts:
```diff
@@ -0,0 +1,74 @@
+      if (!isServiceLoaded) {
+        return new InternalServerErrorException('Minio Client has not loaded')
+      }
+      await minioClient.removeObject(bucketName, fileName)
```
Use try catch

rajdip-b commented on apps/api/src/provider/minio.provider.ts:
Wrap all the three functions in this class with try catch to catch unexpected errors encountered from bucket handling.

rajdip-b commented:
> @rajdip-b i have added error handling and implemented other comments has you said. One thing i was not able to figure out is the error types. i mean `catch (error: <type>)` so we can might handle it in a better way

I don't think you need to catch the types. The type will be very much expected to be caused by bucket errors, so it is fine.

rajdip-b commented on apps/api/src/provider/minio.provider.ts:
```diff
@@ -0,0 +1,110 @@
+    async function createBucketIfNotExists() {
+      if (!minioClient) return
+
+      const bucketExists = await minioClient.bucketExists(bucketName)
```
You left our this line!

HarshPatel5940 replied:
Ouchh... Sorry!!

HarshPatel5940 commented:
> > @rajdip-b i have added error handling and implemented other comments has you said. One thing i was not able to figure out is the error types. i mean `catch (error: <type>)` so we can might handle it in a better way
> 
> I don't think you need to catch the types. The type will be very much expected to be caused by bucket errors, so it is fine.

Oh ok, noted.

rajdip-b reviewed (changes requested):
I see you are returning errors. Throw them instead. Otherwise the code will be a lot more difficult!

HarshPatel5940 commented:
> I see you are returning errors. Throw them instead. Otherwise the code will be a lot more difficult!

are you talking about [line 24](https://github.com/HarshPatel5940/keyshade/blob/ce6d7161bbfa5d0410e632afa5f4aa179dbb9491/apps/api/src/provider/minio.provider.ts#L24)? 

instead of return -> throw ? and is the exit in the place you needed?

rajdip-b commented:
@HarshPatel5940 talking about these:
- https://github.com/HarshPatel5940/keyshade/blob/ce6d7161bbfa5d0410e632afa5f4aa179dbb9491/apps/api/src/provider/minio.provider.ts#L75
- https://github.com/HarshPatel5940/keyshade/blob/ce6d7161bbfa5d0410e632afa5f4aa179dbb9491/apps/api/src/provider/minio.provider.ts#L96
- https://github.com/HarshPatel5940/keyshade/blob/ce6d7161bbfa5d0410e632afa5f4aa179dbb9491/apps/api/src/provider/minio.provider.ts#L111

HarshPatel5940 commented:
> @HarshPatel5940 talking about these:
> 
> * https://github.com/HarshPatel5940/keyshade/blob/ce6d7161bbfa5d0410e632afa5f4aa179dbb9491/apps/api/src/provider/minio.provider.ts#L75
> * https://github.com/HarshPatel5940/keyshade/blob/ce6d7161bbfa5d0410e632afa5f4aa179dbb9491/apps/api/src/provider/minio.provider.ts#L96
> * https://github.com/HarshPatel5940/keyshade/blob/ce6d7161bbfa5d0410e632afa5f4aa179dbb9491/apps/api/src/provider/minio.provider.ts#L111

🤦🏼 my bad. Whoopsie

HarshPatel5940 commented:
Fixed it with https://github.com/keyshade-xyz/keyshade/pull/237/commits/204eab9441cabc658e4e87ce8f28f4b11d0c1486

rajdip-b commented:
:tada: This PR is included in version 1.4.0 :tada:

The release is available on [GitHub release](https://github.com/keyshade-xyz/keyshade/releases/tag/v1.4.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
